### PR TITLE
Generate proper SQLAlchemy classes and tables for self relationships

### PR DESCRIPTION
**Pull request by realVinayak**
See (https://github.com/specify/specify7/issues/2305#issuecomment-1627468923)

### CODE REVIEW

**Review by grantfitzsimmons — APPROVED** [review]
⚠️🚨 **USER REVIEW ONLY!**

It looks like it's working correctly for queries on `sp7demofish` and `fwri`. It very well could break many things, but as a fix for #2305, it seems to work. We need another UX tester in addition to a developer or two. Please read @realVinayak's comment linked at the top of this PR.

File: specifyweb/stored_queries/build_models.py [hunk 1]
```diff
@@ -78,11 +76,11 @@ def make_relationship(reldef):
             column = getattr(table.c, reldef.column)
 
             relationship_args = {'foreign_keys': column}
+            if remote_class is cls:
+                relationship_args['remote_side'] = table.c[ tabledef.idColumn ]
```
Comment (melton-jason) [inline specifyweb/stored_queries/build_models.py:80]: I will have to read up a little more on the sqlalchemy docs to completely understand everything that is going on here, but when compared to how it was previously,  this look alright!

File: specifyweb/stored_queries/build_models.py [hunk 2]
```diff
@@ -78,11 +76,11 @@ def make_relationship(reldef):
...
 
             if hasattr(reldef, 'otherSideName'):
                 backref_args = {'uselist': reldef.type != 'one-to-one'}
-                if remote_class is cls:
-                    backref_args['remote_side'] = table.c[ tabledef.idColumn ]
 
                 relationship_args['backref'] = orm.backref(reldef.otherSideName, **backref_args)
```
Comment (melton-jason) [inline specifyweb/stored_queries/build_models.py:85]: We will have to look at refactoring this code in the future. 
sqlalchemy has "deprecated" `orm.backref` in favor of `relationship.back_populates` explicitly using `relationship` objects. 
https://docs.sqlalchemy.org/en/20/orm/backref.html

I'm sure there are plenty of other things like this as well that we should address.

**Comment by grantfitzsimmons** [comment]
@realVinayak Once you feel confident in this, please make it a full PR and request review as you see fit. Can you include testing instructions as well?

**Comment by realVinayak** [comment]
@grantfitzsimmons I'd actually like to take some time to write automated test for it - to check if actual datamodel and SQLAlchemy datamodel are same or not. I'm half way done through that (in fact my tests also spotted more fields being flipped like this in previous versions of SQLAlchemy classes - for example Agent->Organization - which are now fixed in this branch). 
That automated test would take away the need of most UI testing for this - and would also catch errors like this in future (the current code makes some subtle assumptions about the datamodel - if datamodel changes, it would generate wrong relationships).
I'm on track to wrap automated testing by 7.9.1. Does that work?

File: specifyweb/stored_queries/build_models.py (hunk 2 above)
Comment (melton-jason) [reply specifyweb/stored_queries/build_models.py:85]: Ah, yes, no worries then. 
This works for now.

**Comment by specifysoftware** [comment]
This pull request has been mentioned on **Specify Community Forum**. There might be relevant details there:

https://discourse.specifysoftware.org/t/specify-7-9-2-release-announcement/1452/1

**Pull request merged**